File: src/main/php/Utc.php
<?php

declare(strict_types=1);

namespace PetrKnap\ZonedDateTimePersistence;

use DateTimeImmutable;
use DateTimeInterface;
use Doctrine\ORM\Mapping as ORM;
use LogicException;

/**
 * @internal base
 *
 * @phpstan-import-type LocalDateTime from JavaSe8\Time
 * @phpstan-import-type ZonedDateTime from JavaSe8\Time
 */
#[ORM\MappedSuperclass]
abstract class Utc
{
    /**
     * @var LocalDateTime|null
     */
    #[ORM\Column(type: 'datetime_immutable', nullable: true)]
    protected readonly DateTimeImmutable|null $utc;

    protected function __construct(
        DateTimeInterface $zonedDateTime,
    ) {
        $this->utc = ZonedDateTimePersistence::computeUtcDateTime(JavaSe8\Time::zonedDateTime($zonedDateTime));
    }

    public function asNullable(): static|null
    {
        return $this->utc === null ? null : $this;
    }

    /**
     * @return ($format is false ? LocalDateTime : string)
     */
    public function getUtcDateTime(string|false $format = false): DateTimeImmutable|string
    {
        if ($this->utc === null) {
            $this->thisInstanceShouldBeNull();
        }
        return $format ? $this->utc->format($format) : $this->utc;
    }

    /**
     * @return ZonedDateTime
     */
    abstract public function toZonedDateTime(): DateTimeImmutable;

    protected function thisInstanceShouldBeNull(): never
    {
        throw new LogicException(get_class($this) . ' instance was created without data, call asNullable() method or adjust your object–relational mapping');
    }
}

File: src/main/java/io/github/petrknap/persistence/zoneddatetime/Utc.java
package io.github.petrknap.persistence.zoneddatetime;

import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

@MappedSuperclass
abstract class Utc<T extends Utc<T>> {
    @Column(nullable = true)
    private @Nullable LocalDateTime utc;

    protected Utc(@NotNull ZonedDateTime zonedDateTime)
    {
        utc = ZonedDateTimePersistence.computeUtcDateTime(zonedDateTime);
    }

    protected Utc() {}

    public @Nullable T asNullable()
    {
        return utc == null ? null : (T) this;
    }

    public @NotNull LocalDateTime getUtcDateTime()
    {
        if (utc == null) {
            thisInstanceShouldBeNull();
        }
        return utc;
    }

    public @NotNull String getUtcDateTime(@NotNull String format)
    {
        return DateTimeFormatter.ofPattern(format).format(getUtcDateTime());
    }

    public abstract @NotNull ZonedDateTime toZonedDateTime();

    protected void thisInstanceShouldBeNull() throws IllegalStateException
    {
        throw new IllegalStateException(getClass().getName() + " instance was created without data, call asNullable() method or adjust your object–relational mapping");
    }
}
